tests/integration/git_annex_archive_test.go

```diff
@@ -0,0 +1,197 @@
+
+		// get the commitID of the master branch
+		repo, err := git.OpenRepository(git.DefaultContext, remoteRepoPath)
+		require.NoError(t, err)
+		commitID, err := repo.GetBranchCommitID("master")
+		require.NoError(t, err)
+		tree, err := repo.GetTree("master")
+		require.NoError(t, err)
+		entries, err := tree.ListEntriesRecursiveFast()
+		require.NoError(t, err)
+		filesInGit := make([]string, 0, len(entries))
+		for _, entry := range entries {
+			if !entry.IsDir() {
+				filesInGit = append(filesInGit, entry.Name())
+			}
+		}
+
+		t.Run("TARGZ", func(t *testing.T) {
+			// request a tar.gz archive of the repo
+			link, _ := url.Parse(fmt.Sprintf("/api/v1/repos/%s/%s/archive/master.tar.gz", ctx.Username, ctx.Reponame))
+			resp := ctx.Session.MakeRequest(t, NewRequest(t, "GET", link.String()), http.StatusOK)
+			bs, err := io.ReadAll(resp.Body)
+			require.NoError(t, err)
+
+			// open the archive for reading
+			gzrd, err := gzip.NewReader(bytes.NewReader(bs))
+			require.NoError(t, err)
+			defer gzrd.Close()
+			rd := tar.NewReader(gzrd)
+
+			var filesInArchive []string
+			for {
+				header, err := rd.Next()
+				if err == io.EOF {
+					break
+				}
+				require.NoError(t, err)
+
+				// skip directories
+				if header.Typeflag == tar.TypeDir {
+					continue
+				}
+
+				// check that the pax_global_header comment field is correctly set
+				if path.Base(header.Name) == "pax_global_header" {
+					require.Equal(t, commitID, header.PAXRecords["comment"])
+					continue // skip the remaining checks since this file does not exist in git
+				}
+
+				name := strings.TrimPrefix(header.Name, ctx.Reponame+"/")
+				filesInArchive = append(filesInArchive, name)
+
+				blob, err := tree.GetBlobByPath(name)
+				require.NoError(t, err)
+				isAnnexed, err := annex.IsAnnexed(blob)
+				require.NoError(t, err)
+
+				// make sure all files are the same as in the repo itself
+				actualContent, err := io.ReadAll(rd)
+				require.NoError(t, err)
+				if isAnnexed {
+					fa, err := annex.Content(blob)
+					require.NoError(t, err)
+					defer fa.Close()
+
+					// check that the file mode (type and permissions) is equal
+					actualFileMode := header.FileInfo().Mode()
+					stat, err := fa.Stat()
+					require.NoError(t, err)
+					expectedFileMode := stat.Mode()
+					require.Equal(t, expectedFileMode, actualFileMode)
+
+					// check that the content is equal
+					expectedContent, err := io.ReadAll(fa)
+					require.NoError(t, err)
+					require.Equal(t, expectedContent, actualContent)
+				} else {
+					// check that the content is equal
+					r, err := blob.DataAsync()
+					require.NoError(t, err)
+					defer r.Close()
+					expectedContent, err := io.ReadAll(r)
+					require.NoError(t, err)
+					require.Equal(t, expectedContent, actualContent)
+				}
+			}
+			// check that all files that are in git are also present in the archive
+			require.Equal(t, filesInGit, filesInArchive)
```

Comment:
I wonder if you need to sort these first. tarballs (and zipballs) can come in any order and still produce the same output when extracted.

In this case, with only a few files, and go handling both the zipping and unzipping, it's likely but not guaranteed that the order will be the same on both sides.

Reply:
Good catch, those should definitely be sorted first.